internal/trie/node/branch.go

```diff
@@ -0,0 +1,50 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package node
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+var _ Node = (*Branch)(nil)
+
+// Branch is a branch in the trie.
+type Branch struct {
+	Key      []byte // partial key
+	Children [16]Node
+	Value    []byte
+	// dirty is true when the branch differs
+	// from the node stored in the database.
+	dirty      bool
+	hashDigest []byte
+	encoding   []byte
+	// generation is incremented on every trie Snapshot() call.
+	// Nodes that are part of the trie are then gradually updated
+	// to have a matching generation number as well, if they are
+	// still relevant.
+	generation uint64
```

Comment:
not sure what this part of the comment means, I'd just remove it haha

Reply:
the generation is used to track distinct "versions" of the trie, which may point to nodes from previous generations. eg, when we sync a new block, we make a snapshot, and then all new nodes that are added during the state transition process have the new generation, but the old unmodified nodes have the older generation

Reply:
Updated in 8c3178891f3f0aeeb842482f8c9213619e04b232